macros/contexts/contextSignificantFigures.pl

```diff
@@ -136,72 +131,62 @@ sub new {
 # Either return the current number of sigfigs for the number or set the current number.
 
 sub sigfigs {
-	my ($self, $num_sigfigs) = @_;
-	return $self->{sigfigs} unless defined($num_sigfigs);
-	Value::Error('The number of significant figures must be inf or an integer >=-1')
-		unless $num_sigfigs =~ m/^inf$/ || ($num_sigfigs =~ /-?\d+/ && $num_sigfigs >= 1);
-	my ($v, $exp);
-	if ($num_sigfigs =~ m/^inf$/) {
-		$v   = $self->value;
-		$exp = 0;
-	} else {
-		($v, $exp) = split(/E/, $self->value);
-		$exp = 0 unless defined($exp);    # if $self->value is not in exponential form.
-	}
-
-	# Calculate the digit to round to.
-	if ($num_sigfigs !~ /^inf$/) {
-		my $N = $num_sigfigs - 1;
-		$self->{data} = [ sprintf("%.${N}E", $v * 10**$exp) ];
-	} else {
-		$self->{data} = [$v];
-	}
-
-	$self->{sigfigs} = $num_sigfigs;
+	my ($self, $n) = @_;
+	return $self->{sigfigs} if !defined($n) || $self->{sigfigs} == $n;
+	$self->{sigfigs} = $self->context->checkSigFigs($n);
+	$self->{data}[0] = $n == 'inf' ? $self->value : $self->format('E', ROUND($self->value, $n), $n);
 	return $self->{sigfigs};
 }
 
-# return the exponent since stored in the form d * 10^(p), where d is an integer.
-sub exp { return shift->{exp}; }
+# Shortcut for returning the number of significant figures.
+sub N { shift->{sigfigs} }
 
-# return the mantissa of the number (without sign in [1,10) )
-sub mantissa { return (split(/E/, shift->value))[0] + 0; }
+# Return the exponent.
+sub E { shift->{exp} }
 
-#  Stringify and TeXify the number in the context's base
-sub string {
-	my $self = shift;
-	return "" . $self->value if $self->sigfigs =~ /inf/;
-	my $N = -$self->{exp} + $self->sigfigs - 1;     # digit to round to.
-	my $v = $self->mantissa * 10**($self->{exp});
-	return abs($self->{exp}) > 5 ? $self->value : ($N > 0 ? sprintf("%.${N}f", $v) : "$v");
+# Return the exponential for the given $value with max(3,14-$exp) sigfigs.  This is
+# helpful for addition and subtraction.
+sub expFor {
+	my ($self, $value, $exp) = @_;
+	$exp = main::max(3, 14 - $exp);
+	return (split(/E/, $self->format('E', $value, $exp)))[1] + 0;
```

Comment:
I suppose it would be better to use `$n` rather than `$exp` in these lines, as that is actually what is being computed.

Also, line 151 should probably be

``` perl
	$exp = main::max($exp, 14 - $exp);
```

(or with `$n` instead of `$exp`), as it is in the `ROUND()` function.

Those are both my fault, of course.